cpp/src/arrow/csv/csv-to-arrow.cc

```diff
@@ -0,0 +1,162 @@
+DEFINE_string(escape_char, "\\", "Escaping character (if `escaping` is true)");
+DEFINE_bool(newlines_in_values, false,
+            "Values are allowed to contain CR (0x0d) and LF (0x0a) characters");
+DEFINE_bool(ignore_empty_lines, true, "Ignore empty lines");
+DEFINE_int32(
+    header_rows, 1,
+    "Number of header rows to skip (including the first row containing column names)");
+
+DEFINE_validator(delimiter, &CheckCharacter);
+DEFINE_validator(quote_char, &CheckCharacter);
+DEFINE_validator(escape_char, &CheckCharacter);
+DEFINE_validator(header_rows, &CheckPositive);
+
+// Conversion options
+DEFINE_bool(check_utf8, true, "Check UTF8 validity of string columns");
+
+// Read options
+DEFINE_bool(use_threads, true, "Use the global CPU thread pool");
+DEFINE_int32(block_size, 1 << 20, "Block size");
+DEFINE_validator(block_size, &CheckPositive);
+
+// Tool options
+DEFINE_bool(verbose, true, "Verbose output");
+
+namespace arrow {
+namespace csv {
+
+static Status InitParseOptions(ParseOptions& options) {
+  options.delimiter = FLAGS_delimiter.at(0);
+  options.quoting = FLAGS_quoting;
+  options.quote_char = FLAGS_quote_char.at(0);
+  options.double_quote = FLAGS_double_quote;
+  options.escaping = FLAGS_escaping;
+  options.escape_char = FLAGS_escape_char.at(0);
+  options.newlines_in_values = FLAGS_newlines_in_values;
+  options.ignore_empty_lines = FLAGS_ignore_empty_lines;
+  options.header_rows = FLAGS_header_rows;
+  return Status::OK();
+}
+
+static Status InitConvertOptions(ConvertOptions& options) {
+  options.check_utf8 = FLAGS_check_utf8;
+  return Status::OK();
+}
+
+static Status InitReadOptions(ReadOptions& options) {
+  options.use_threads = FLAGS_use_threads;
+  options.block_size = FLAGS_block_size;
+  return Status::OK();
+}
+
+static Status Run(int argc, char** argv) {
+  // Instantiate and handle options
+  gflags::SetUsageMessage("csv-to-arrow [FLAGS] input.csv  > output.arrow");
+  gflags::ParseCommandLineFlags(&argc, &argv, true);
+  if (argc != 2) {
+    std::cerr << "Usage: " << gflags::ProgramUsage() << std::endl;
+    return Status::Invalid("Input filename missing");
+  }
+  auto parse_options = ParseOptions::Defaults();
+  ARROW_RETURN_NOT_OK(InitParseOptions(parse_options));
+  auto read_options = ReadOptions::Defaults();
+  ARROW_RETURN_NOT_OK(InitReadOptions(read_options));
+  auto convert_options = ConvertOptions::Defaults();
+  ARROW_RETURN_NOT_OK(InitConvertOptions(convert_options));
+
+  // Instantiate reading
+  std::string input_filename(argv[1]);
+  std::shared_ptr<io::ReadableFile> input_file;
+  ARROW_RETURN_NOT_OK(io::ReadableFile::Open(input_filename, &input_file));
+  auto pool = default_memory_pool();
+  std::shared_ptr<TableReader> reader;
+  ARROW_RETURN_NOT_OK(TableReader::Make(pool, input_file, read_options, parse_options,
+                                        convert_options, &reader));
+
+  // Instantiate writing, read from input and write to output
+  io::StdoutStream stdout;
```

Comment:
I'm not sure that's a good idea. `io::StdoutStream` uses the C++ `std::cout` stream.
1) it might be slow with large data
2) it might not be open in binary mode on Windows

In any case, since the output is binary, you don't gain much by being able to displa it on a terminal. The command-line should probably take an output filename instead.

Reply:
I actually hadn't considered 2, so thanks for pointing that out.

The original motivation here was to pipe to [`stream-to-file.cc`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/stream-to-file.cc) or redirect directly to a file, and I think I can just revert to the prior commit: https://github.com/apache/arrow/pull/3136/commits/3f259d55e147e86d64db398864c89a6ce186d8d3